### PR TITLE
Add code to schedule GEMMs

### PR DESCRIPTION
**Pull request by harsh-nod**
This PR adds the code to schedule tkw gemms. We
have to construct edges for the modulo scheduler
and define resources that are available for the
scheduler. These resources are tunable.

### CODE REVIEW

**Comment by harsh-nod** [comment]
Some sample images for the schedule in the test.

scheduling_fx_graph

**Comment by harsh-nod** [comment]
<html>
<table id="T_71968">
  <thead>
    <tr>
      <th class="blank level0" >&nbsp;</th>
      <th id="T_71968_level0_col0" class="col_heading level0 col0" >Stage 0</th>
      <th id="T_71968_level0_col1" class="col_heading level0 col1" >Stage 1</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th id="T_71968_level0_row0" class="row_heading level0 row0" >0</th>
      <td id="T_71968_row0_col0" class="data row0 col0" >read_5<br>read_4<br></td>
      <td id="T_71968_row0_col1" class="data row0 col1" ></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row1" class="row_heading level0 row1" >1</th>
      <td id="T_71968_row1_col0" class="data row1 col0" ></td>
      <td id="T_71968_row1_col1" class="data row1 col1" ></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row2" class="row_heading level0 row2" >2</th>
      <td id="T_71968_row2_col0" class="data row2 col0" ></td>
      <td id="T_71968_row2_col1" class="data row2 col1" ></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row3" class="row_heading level0 row3" >3</th>
      <td id="T_71968_row3_col0" class="data row3 col0" ></td>
      <td id="T_71968_row3_col1" class="data row3 col1" ></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row4" class="row_heading level0 row4" >4</th>
      <td id="T_71968_row4_col0" class="data row4 col0" ></td>
      <td id="T_71968_row4_col1" class="data row4 col1" ></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row5" class="row_heading level0 row5" >5</th>
      <td id="T_71968_row5_col0" class="data row5 col0" >write_3<br>write_2<br></td>
      <td id="T_71968_row5_col1" class="data row5 col1" ></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row6" class="row_heading level0 row6" >6</th>
      <td id="T_71968_row6_col0" class="data row6 col0" >read_shared_0_1_1<br>read_shared_0_1_0<br></td>
      <td id="T_71968_row6_col1" class="data row6 col1" ></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row7" class="row_heading level0 row7" >7</th>
      <td id="T_71968_row7_col0" class="data row7 col0" >read_shared_0_0_1<br>read_shared_0_0_3<br></td>
      <td id="T_71968_row7_col1" class="data row7 col1" ></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row8" class="row_heading level0 row8" >8</th>
      <td id="T_71968_row8_col0" class="data row8 col0" >read_shared_0_0_0<br>read_shared_0_0_2<br></td>
      <td id="T_71968_row8_col1" class="data row8 col1" ></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row9" class="row_heading level0 row9" >9</th>
      <td id="T_71968_row9_col0" class="data row9 col0" >read_shared_1_0_1<br>read_shared_1_0_0<br></td>
      <td id="T_71968_row9_col1" class="data row9 col1" ></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row10" class="row_heading level0 row10" >10</th>
      <td id="T_71968_row10_col0" class="data row10 col0" >mma_0_0_0<br>mma_1_0_0<br></td>
      <td id="T_71968_row10_col1" class="data row10 col1" ></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row11" class="row_heading level0 row11" >11</th>
      <td id="T_71968_row11_col0" class="data row11 col0" >mma_1_1_0<br>mma_0_1_0<br></td>
      <td id="T_71968_row11_col1" class="data row11 col1" >read_5<br>read_4<br></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row12" class="row_heading level0 row12" >12</th>
      <td id="T_71968_row12_col0" class="data row12 col0" >mma_0_0_1<br>mma_1_0_1<br></td>
      <td id="T_71968_row12_col1" class="data row12 col1" ></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row13" class="row_heading level0 row13" >13</th>
      <td id="T_71968_row13_col0" class="data row13 col0" >mma_1_1_1<br>mma_0_1_1<br></td>
      <td id="T_71968_row13_col1" class="data row13 col1" ></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row14" class="row_heading level0 row14" >14</th>
      <td id="T_71968_row14_col0" class="data row14 col0" >acc_0_0_0<br>acc_1_0_0<br></td>
      <td id="T_71968_row14_col1" class="data row14 col1" ></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row15" class="row_heading level0 row15" >15</th>
      <td id="T_71968_row15_col0" class="data row15 col0" >acc_1_1_0<br>acc_0_1_0<br></td>
      <td id="T_71968_row15_col1" class="data row15 col1" ></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row16" class="row_heading level0 row16" >16</th>
      <td id="T_71968_row16_col0" class="data row16 col0" ></td>
      <td id="T_71968_row16_col1" class="data row16 col1" >write_3<br>write_2<br></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row17" class="row_heading level0 row17" >17</th>
      <td id="T_71968_row17_col0" class="data row17 col0" ></td>
      <td id="T_71968_row17_col1" class="data row17 col1" >read_shared_0_1_1<br>read_shared_0_1_0<br></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row18" class="row_heading level0 row18" >18</th>
      <td id="T_71968_row18_col0" class="data row18 col0" ></td>
      <td id="T_71968_row18_col1" class="data row18 col1" >read_shared_0_0_1<br>read_shared_0_0_3<br></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row19" class="row_heading level0 row19" >19</th>
      <td id="T_71968_row19_col0" class="data row19 col0" ></td>
      <td id="T_71968_row19_col1" class="data row19 col1" >read_shared_0_0_0<br>read_shared_0_0_2<br></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row20" class="row_heading level0 row20" >20</th>
      <td id="T_71968_row20_col0" class="data row20 col0" ></td>
      <td id="T_71968_row20_col1" class="data row20 col1" >read_shared_1_0_1<br>read_shared_1_0_0<br></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row21" class="row_heading level0 row21" >21</th>
      <td id="T_71968_row21_col0" class="data row21 col0" ></td>
      <td id="T_71968_row21_col1" class="data row21 col1" >mma_0_0_0<br>mma_1_0_0<br></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row22" class="row_heading level0 row22" >22</th>
      <td id="T_71968_row22_col0" class="data row22 col0" ></td>
      <td id="T_71968_row22_col1" class="data row22 col1" >mma_1_1_0<br>mma_0_1_0<br></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row23" class="row_heading level0 row23" >23</th>
      <td id="T_71968_row23_col0" class="data row23 col0" ></td>
      <td id="T_71968_row23_col1" class="data row23 col1" >mma_0_0_1<br>mma_1_0_1<br></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row24" class="row_heading level0 row24" >24</th>
      <td id="T_71968_row24_col0" class="data row24 col0" ></td>
      <td id="T_71968_row24_col1" class="data row24 col1" >mma_1_1_1<br>mma_0_1_1<br></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row25" class="row_heading level0 row25" >25</th>
      <td id="T_71968_row25_col0" class="data row25 col0" ></td>
      <td id="T_71968_row25_col1" class="data row25 col1" >acc_0_0_0<br>acc_1_0_0<br></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row26" class="row_heading level0 row26" >26</th>
      <td id="T_71968_row26_col0" class="data row26 col0" ></td>
      <td id="T_71968_row26_col1" class="data row26 col1" >acc_1_1_0<br>acc_0_1_0<br></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row27" class="row_heading level0 row27" >27</th>
      <td id="T_71968_row27_col0" class="data row27 col0" ></td>
      <td id="T_71968_row27_col1" class="data row27 col1" ></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row28" class="row_heading level0 row28" >28</th>
      <td id="T_71968_row28_col0" class="data row28 col0" ></td>
      <td id="T_71968_row28_col1" class="data row28 col1" ></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row29" class="row_heading level0 row29" >29</th>
      <td id="T_71968_row29_col0" class="data row29 col0" ></td>
      <td id="T_71968_row29_col1" class="data row29 col1" ></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row30" class="row_heading level0 row30" >30</th>
      <td id="T_71968_row30_col0" class="data row30 col0" ></td>
      <td id="T_71968_row30_col1" class="data row30 col1" ></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row31" class="row_heading level0 row31" >31</th>
      <td id="T_71968_row31_col0" class="data row31 col0" ></td>
      <td id="T_71968_row31_col1" class="data row31 col1" ></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row32" class="row_heading level0 row32" >32</th>
      <td id="T_71968_row32_col0" class="data row32 col0" ></td>
      <td id="T_71968_row32_col1" class="data row32 col1" ></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row33" class="row_heading level0 row33" >33</th>
      <td id="T_71968_row33_col0" class="data row33 col0" ></td>
      <td id="T_71968_row33_col1" class="data row33 col1" ></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row34" class="row_heading level0 row34" >34</th>
      <td id="T_71968_row34_col0" class="data row34 col0" ></td>
      <td id="T_71968_row34_col1" class="data row34 col1" ></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row35" class="row_heading level0 row35" >35</th>
      <td id="T_71968_row35_col0" class="data row35 col0" ></td>
      <td id="T_71968_row35_col1" class="data row35 col1" ></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row36" class="row_heading level0 row36" >36</th>
      <td id="T_71968_row36_col0" class="data row36 col0" ></td>
      <td id="T_71968_row36_col1" class="data row36 col1" ></td>
    </tr>
    <tr>
      <th id="T_71968_level0_row37" class="row_heading level0 row37" >37</th>
      <td id="T_71968_row37_col0" class="data row37 col0" ></td>
      <td id="T_71968_row37_col1" class="data row37 col1" ></td>
    </tr>
  </tbody>
</table>
</html>

**Comment by harsh-nod** [comment]
The html file looks like above, but the kernel is highlighted green.

**Comment by harsh-nod** [comment]
This is what the schedule looks like.
Screenshot 2024-09-11 at 6 00 24 PM

File: shark_turbine/kernel/wave/visualization.py [hunk 1]
```diff
@@ -3,7 +3,14 @@
     import pygraphviz as pgv
 except:
     graphviz_disabled = True
+pandas_disabled = False
+try:
+    import pandas as pd
+except:
+    matplotlib_disabled = True
```
Comment (Hardcode84) [inline shark_turbine/kernel/wave/visualization.py:10]: `pandas_disabled`

**Review by Hardcode84 — APPROVED** [review]
LGTM modulo couple comments

File: shark_turbine/kernel/wave/visualization.py (hunk 1 above)
Comment (harsh-nod) [reply shark_turbine/kernel/wave/visualization.py:10]: thanks will change.

File: shark_turbine/kernel/wave/visualization.py [hunk 2]
```diff
@@ -22,3 +29,69 @@ def visualize_graph(graph: fx.Graph, file_name: str):
             G.add_edge(node_numbering[id(node)], node_numbering[id(user)])
     G.layout(prog="dot")
     G.draw(file_name)
+
+
+def visualize_edges(edges: list[Edge], file_name: str):
+    if graphviz_disabled:
```
[the remark this answers is not in the document]
Comment (harsh-nod) [reply shark_turbine/kernel/wave/visualization.py:35]: sure I will do this for the pandas import. The graphviz import is used in multiple functions so will leave that as is.

**Pull request merged**